substrate-archive/src/database/batch.rs

```diff
@@ -147,14 +158,42 @@ impl Batch {
 		if self.len > 0 {
 			for mut chunk in self.chunks {
 				chunk.append(&self.trailing);
-				let done = chunk.execute(conn).await?;
+				let done = chunk.execute(&mut *conn).await?;
 				rows_affected += done;
 			}
 		}
 
 		Ok(rows_affected)
 	}
 
+	/// Execute a batch concurrently. If a batch size is greater than CHUNK_MAX,
+	/// the query will be executed in multiple independent futures.
+	/// The amount of futures is controlled with `limit`. A limit of `None`
+	/// will limit this function to the number of idle database connections.
+	pub async fn execute_concurrent(self, conn: PgPool, limit: Option<usize>) -> Result<u64> {
+		let rows_affected = AtomicUsize::new(0);
+		let trailing = self.trailing.clone();
+		let num_idle = |_| async {
+			let c = conn.clone();
+			Some(timeout(Duration::from_millis(100), task::spawn_blocking(move || c.num_idle())).await).transpose()
```

Comment:
Yes, I was a bit worried about that as well. In my experience running the archive, it hasn't been an issue, but it's there just in case things stall. In the worst case, we wait 100ms and then have a longer insert for only this batch because of the hard-coded 2 connection limit if we can't find idle connections.